code/core/baseHandler.py

```diff
@@ -292,31 +341,44 @@ async def route_query_based_on_tools(self):
             return
 
         top_tool = self.tool_routing_results[0] 
-        tool_name = top_tool['tool'].name
+        tool = top_tool['tool']
+        tool_name = tool.name
+        params = top_tool['result']
         log(f"Selected tool: {tool_name}")
         
-        if tool_name == "search":
-            logger.info("Routing to search functionality")
-            await self.get_ranked_answers()
-        elif tool_name == "details":
-            logger.info("Routing to item details functionality")
-            params = top_tool['result']
-            await item_details.ItemDetailsHandler(params, self).do()
-        elif tool_name == "compare":
-            logger.info("Routing to comparison functionality")
-            params = top_tool['result']
-            await compare_items.CompareItemsHandler(params, self).do()
-        elif tool_name == "accompaniment":
-            logger.info("Routing to accompaniment functionality")
-            params = top_tool['result']
-            await accompaniment.AccompanimentHandler(params, self).do()
-        elif tool_name == "recipe_substitutions":
-            logger.info("Routing to substitution functionality")
-            params = top_tool['result']
-            await substitution.SubstitutionHandler(params, self).do()
+        # Check if tool has a handler class defined
+        if tool.handler_class:
+            try:
+                logger.info(f"Routing to {tool_name} functionality via {tool.handler_class}")
+                
+                # For non-search tools, clear any items that FastTrack might have populated
+                if tool_name != "search":
+                    self.final_retrieved_items = []
+                    self.retrieved_items = []
+                
+                # Dynamic import of handler module and class
+                module_path, class_name = tool.handler_class.rsplit('.', 1)
+                module = importlib.import_module(module_path)
+                handler_class = getattr(module, class_name)
+                
```

Comment:
If we haven't already, we should document that your environment may be dynamically modified.  If people forget to create a virtual environment, they could run into issues.